src/mlpack/tests/main_tests/krann_test.cpp

```diff
@@ -0,0 +1,335 @@
+}
+ 
+/*
+ * Check that we can't pass an invalid value of tau.
+ */
+BOOST_AUTO_TEST_CASE(KRANNInvalidTauTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of  top tau %ile reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 5);
+  SetInputParam("tau", (double) -1); // Invalid.
+
+  Log::Fatal.ignoreInput = true;
+  BOOST_REQUIRE_THROW(mlpackMain(), std::runtime_error);
+  Log::Fatal.ignoreInput = false;
+}
+
+/**
+ * Make sure that dimensions of the neighbors and distances matrices are correct
+ * given a value of k.
+ */
+BOOST_AUTO_TEST_CASE(KRANNOutputDimensionTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  // Random input, some k <= number of  top tau %ile reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("k", (int) 5);
+
+  mlpackMain();
+
+  // Check the neighbors matrix has 5 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>
+      ("neighbors").n_rows, 5);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::Mat<size_t>>
+      ("neighbors").n_cols, 100);
+
+  // Check the distances matrix has 10 points for each input point.
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("distances").n_rows, 5);
+  BOOST_REQUIRE_EQUAL(CLI::GetParam<arma::mat>("distances").n_cols, 100);
+}
+
+/**
+ * Ensure that saved model can be used again.
+ */
+BOOST_AUTO_TEST_CASE(KRANNModelReuseTest)
+{
+  arma::mat referenceData;
+  referenceData.randu(3, 100); // 100 points in 3 dimensions.
+
+  arma::mat queryData;
+  queryData.randu(3, 90); // 90 points in 3 dimensions.
+
+  // Random input, some k <= number of  top tau %ile reference points.
+  SetInputParam("reference", std::move(referenceData));
+  SetInputParam("query", queryData);
+  SetInputParam("k", (int) 5);
+
+  mlpackMain();
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  RANNModel* output_model;
+  neighbors = std::move(CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  distances = std::move(CLI::GetParam<arma::mat>("distances"));
+  output_model = std::move(CLI::GetParam<RANNModel*>("output_model"));
+
+  // Reset passed parameters.
+  CLI::GetSingleton().Parameters()["reference"].wasPassed = false;
+  CLI::GetSingleton().Parameters()["query"].wasPassed = false;
+
+  // Input saved model, pass the same query and keep k unchanged.
+  SetInputParam("input_model", output_model);
+  SetInputParam("query", queryData);
+
+  mlpackMain();
+
+  // Check that initial output matrices and the output matrices using
+  // saved model are equal.
+  CheckMatrices(neighbors, CLI::GetParam<arma::Mat<size_t>>("neighbors"));
+  CheckMatrices(distances, CLI::GetParam<arma::mat>("distances"));
+}
+
+/**
+  * Ensure that different leaf sizes give different results.
+ */
+BOOST_AUTO_TEST_CASE(KRANNDifferentLeafSizes)
```

Comment:
Do you think you could add a test like this but for the `tau`, `alpha`, and `tree_type` parameters?  It might also be nice to try a test like this with the `sample_at_leaves` and `single_sample_limit` parameters, but it may be hard to produce a test case where those things actually give different results; so I would say if you're willing, give it a shot and see if the test works, but if the test seems to need a lot of debugging, maybe it's not worth it. :+1:

Reply:
@rcurtin I haven't tried the test cases for `sample_at_leaves` and `single_sample_limit` parameters but I want to include them in the tests. It might take some time but I have added the tests for `tau`, `alpha` and `tree_type` if you could please take a look.

Reply:
Nice, those are actually way nicer than I thought!  I had imagined that for testing you would need to compute neighbors from both models and then check that they are different.  But accessing the values of the parameters directly is *much* simpler and ensures that the parameter is being correctly passed to the models.  It *might* still be nice to do that (just to make sure that those parameters are actually used in the KRANN implementation), but, up to you.  If you can apply the same strategy for `sample_at_leaves` and `single_sample_limit`, I think it could be pretty easy.

Reply:
@rcurtin I was working on the tests when I think I broke something because now when I run any mlpack commands, it gives weird hdf5 errors (for eg. on running any test, `bin/mlpack_test: symbol lookup error: bin/mlpack_test: undefined symbol: arma_H5T_NATIVE_DOUBLE`). I am trying to resolve it. There is one thing I saw that `single_sample_limit` is being considered as `double` which should be `int`.
```
  if (CLI::HasParam("single_sample_limit"))
    rann->SingleSampleLimit() = CLI::GetParam<double>("single_sample_limit");
```
I replaced `double` in the above line by `int` and that was when the hdf5 errors started appearing. 
I only committed the two style changes, not really sure why it is failing the static code analysis check.

Reply:
>  I had imagined that for testing you would need to compute neighbors from both models and then check that they are different.
@rcurtin, can I add those later? We have our exams starting from the second week of July. I would have done it but got stuck with those armadillo related errors.

Reply:
My guess is that for the HDF5 issues, perhaps a package was upgraded on your system or something.  I would try removing your build directory entirely and building again from scratch.  That should hopefully help.

I don't think that it's necessary to update the tests here; when I wrote about the other strategy for testing, I didn't mean that's what we had to do, I just meant that that's the way I had thought to solve it, but I think the approach you've used is just fine. :+1: